com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/verifier/JarVerifier.java

```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    private JarFile openJarFile(URL jarUrl) throws IOException {
+        jarUrl = toJarUrl(jarUrl);
+        JarURLConnection connection = (JarURLConnection) jarUrl.openConnection();
+        return connection.getJarFile();
+    }
+
+    private URL toJarUrl(URL jarUrl) throws MalformedURLException {
+        return new URL("jar:" + jarUrl.toString() + "!/");
+    }
+
+    private List<JarEntry> getJarEntries(JarFile jarFile) throws IOException {
+        List<JarEntry> jarEntries = new ArrayList<>();
+        Enumeration<JarEntry> entries = jarFile.entries();
+        while (entries.hasMoreElements()) {
+            JarEntry jarEntry = readJarEntry(jarFile, entries);
+            if (!jarEntry.isDirectory()) {
+                jarEntries.add(jarEntry);
+            }
+        }
+        return jarEntries;
+    }
+
+    private JarEntry readJarEntry(JarFile jarFile, Enumeration<JarEntry> entries) throws IOException {
+        JarEntry jarEntry = entries.nextElement();
+        FileUtils.validatePath(jarEntry.getName());
+        if (jarEntry.isDirectory()) {
+            return jarEntry;
+        }
+        verifySignature(jarFile, jarEntry);
+        return jarEntry;
+    }
+
+    private void verifySignature(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        try (InputStream inputStream = jarFile.getInputStream(jarEntry)) {
+            while (inputStream.read(buffer, 0, buffer.length) != -1) {
+                // Just read. This will throw a SecurityException if a signature/digest check fails.
+            }
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCertificates, List<JarEntry> jarEntries, String certificateCN) {
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                verifyIsMetaInformation(jarEntry);
+                continue;
+            }
+            verifyJarEntryIsSigned(targetCertificates, entryCertificates, certificateCN);
+        }
+    }
+
+    private void verifyIsMetaInformation(JarEntry jarEntry) {
+        if (!jarEntry.getName()
+                     .startsWith(META_INF)) {
+            throw new SLException(Messages.ARCHIVE_CONTAINS_UNSIGNED_FILES);
+        }
+    }
+
+    private void verifyJarEntryIsSigned(List<X509Certificate> targetCertificates, Certificate[] entryCertificates, String certificateCN) {
+        List<X509Certificate> entryCertificatesList = convertToX509Certificates(entryCertificates);
+        if (!validCertificateChain(targetCertificates, entryCertificatesList, certificateCN)) {
+            throw new SecurityException(Messages.ARCHIVE_IS_NOT_SIGNED_BY_TRUSTED_SIGNER);
+        }
+    }
+
+    private List<X509Certificate> convertToX509Certificates(Certificate[] entryCertificates) {
```

Comment:
Rename `entryCertificates` to `certificates`. This method does not need to know that these certificates belong to a JAR entry. Also, rename to method to `toX509Certificates` for consistency with `toJarUrl`.

Reply:
Done